K8s executor 2

grantnicholas opened the pull request:
I opened this new PR because I couldn't find a way to keep the comments from the old PR while changing the base branch it was merging from. Sorry about that.

dimberman commented on airflow/contrib/kubernetes/pod.py:
```diff
@@ -37,22 +37,27 @@ def __init__(
+        self.args = args or []
+        self.secrets = secrets or []
         self.result = result
-        self.labels = labels if labels else []
```
Good catch :).

dimberman commented on airflow/contrib/executors/kubernetes_executor.py:
```diff
@@ -11,47 +11,121 @@
+    def __init__(self):
+        self.dags_folder = configuration.get('core', 'dags_folder')
+        self.parallelism = configuration.getint('core', 'PARALLELISM')
+        self.kube_image = configuration.get('core', 'k8s_image')
```
kubernetes specific settings should be in the [kubernetes] section.

grantnicholas replied:
+1 on this, this was just me being lazy during development. will fix.

dimberman commented on airflow/contrib/executors/kubernetes_executor.py:
```diff
@@ -11,47 +11,121 @@
+        self.dags_folder = configuration.get('core', 'dags_folder')
+        self.parallelism = configuration.getint('core', 'PARALLELISM')
+        self.kube_image = configuration.get('core', 'k8s_image')
+        self.kube_namespace = self.safe_get('core', 'k8s_namespace', 'default')
```
The namespace is something that should be determined in the creation of the pod. Not the airflow.cfg. https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/

grantnicholas replied:
Gotcha, I was wondering what that env variable was for. Makes sense.

dimberman commented on airflow/contrib/executors/kubernetes_executor.py:
```diff
@@ -11,47 +11,121 @@
+        self.kube_namespace = self.safe_get('core', 'k8s_namespace', 'default')
+
+        # These two props must be set together
+        self.git_repo = self.safe_get('core', 'k8s_git_repo', None)
```
these should also be in the kubernetes section

dimberman commented on airflow/contrib/executors/kubernetes_executor.py:
```diff
@@ -11,47 +11,121 @@
-    :param command:
 
-    :return:
+class PodMaker:
```
Why is the PodMaker in the Executor class? This should be a seperate class so it can be used by the operator as well.

grantnicholas replied:
It is very tightly-coupled to the executor implementation (ie: look at airflow config properties and mount these airflow-dags volumes, otherwise do the git mode pulling). I'm just not sure how reuseable this component is by others.

dimberman replied:
I think that if that's the case you can either inject the executor specific functions or extend the general pod_maker class. The make_pod function definitely would have value elsewhere.

dimberman commented on airflow/contrib/executors/kubernetes_executor.py:
```diff
@@ -115,109 +181,120 @@ def run_next(self, next_job):
-            namespace=self.namespace
+        pod_id = self._create_job_id_from_key(key=key)
+        self.logger.info("k8s: launching image {}".format(self.kube_config.kube_image))
+        pod = self.pod_maker.make_pod(
```
Good call moving a lot of this logic into the pod_maker. That said, if we're no longer going to have unique IDs for every pod we launch we need to be ABSOLUTELY certain that there's no chance of overlap and that we delete all related pods upon starting the airflow instance

grantnicholas replied:
You are correct, and I made a bad bug in this implementation (only using dag_id+task_id as the pod name and these are not guaranteed to be unique). 

There is a more general problem, which is airflow limits the size of `dag_id` and `task_id` to a  max length of 250 and valid pod names max out at 63 (see here:
https://stackoverflow.com/questions/27791913/what-characters-are-allowed-in-kubernetes-port-and-container-names). Thus we can't use `dag_id` or `task_id` in the pod name (ugh). 

Thus, I'm going to go with a uuid approach for the pod name. We still have the labels to identify the task from the pod anyways (dag_id/task_id/execution_date).

dimberman replied:
oof.... I'm not sure how I feel about having uids for pod ids. Why don't we have some truncation logic s.t. if the dag_id + task_id are too long, it only uses the first 25 of the dag_id, the first 25 of the task_id, and then a uuid.

grantnicholas replied:
This is the valid `dag_id/task_id` regex from airflow: `r'^[A-Za-z0-9_\-\.]+$'`. Max len 250.

This is the valid `pod_name` regex from kubernetes: `[a-z0-9]([-a-z0-9]*[a-z0-9])?` Max len 63

It's trickier than you think. We would have to normalize and truncate all the invalid chars. :x:

dimberman commented on airflow/contrib/executors/kubernetes_executor.py:
```diff
@@ -226,36 +303,25 @@ def sync(self):
-            # self.kub_client.delete_job(key)
-            self.logger.info("current running {}".format(self.running))
-            self.running.pop(key)
+            self.kub_client.delete_pod(key)
```
place a comment about why you're deleting pods. Another possibility would be to have a "don't delete pods" config in the airflow.cfg.

grantnicholas replied:
Deleting pods is nice because:

1. Stopped pods/docker containers fill up disk space at an alarmingly fast rate
2. Deleting pods removes those events from kubernetes history, which helps the kubernetes watcher with the "replaying all history" issue. 

I totally agree that a "don't delete pods" config option would be super useful for debugging errors, so I will add that in.

dimberman replied:
I agree that deleting the pods should be the default.

grantnicholas commented:
@dimberman made the changes you suggested outside of the `PodMaker` changes. I still think it's a bit early to separate out this abstraction. If we want to expose something like the `PodMaker` in the public API then I think we can separate it out at a later time.

dimberman commented:
Looks great, Grant! Lemme do some testing/a second run through.

grantnicholas commented:
I found a possible bug. The `KubernetesJobWatcher` process appears to sometimes die randomly and then never start up again. I think we need to monitor the status of this Process and restart it if it fails.

dimberman commented:
@grantnicholas Oof, good catch. Do you have the error logs for the death or is it silent since it's a separate process?

grantnicholas commented:
Sadly silent errors. These are the logs from the last entries I saw that had the "Event" log from the `KubernetesJobWatcher`. 

```
[2017-09-08 21:17:32,613] {kubernetes_executor.py:238} INFO - Attempting to finish pod; pod_id: examplepythonoperatorfromgitsyncsleepfor5-8792299a9e164ad08e381109e5052c19; state: None; labels: {u'airflow-slave': '', u'execution_date': '2017-09-08T14_08_00', u'dag_id': 'example_python_operator_from_git_sync', u'task_id': 'sleep_for_5'}
[2017-09-08 21:17:32,614] {kubernetes_executor.py:241} INFO - finishing job ('example_python_operator_from_git_sync', 'sleep_for_5', datetime.datetime(2017, 9, 8, 14, 8))
[2017-09-08 21:17:32,614] {kubernetes_executor.py:238} INFO - Attempting to finish pod; pod_id: examplepythonoperatorfromgitsyncsleepfor1-51d8fd1729d74b95a89d4a8ad38f418d; state: None; labels: {u'airflow-slave': '', u'execution_date': '2017-09-08T14_08_00', u'dag_id': 'example_python_operator_from_git_sync', u'task_id': 'sleep_for_1'}
[2017-09-08 21:17:32,616] {kubernetes_executor.py:241} INFO - finishing job ('example_python_operator_from_git_sync', 'sleep_for_1', datetime.datetime(2017, 9, 8, 14, 8))
[2017-09-08 21:17:32,616] {kubernetes_executor.py:348} INFO - reporting (('example_python_operator_from_git_sync', 'sleep_for_5', datetime.datetime(2017, 9, 8, 14, 8)), None, 'examplepythonoperatorfromgitsyncsleepfor5-8792299a9e164ad08e381109e5052c19')
[2017-09-08 21:17:32,617] {kubernetes_executor.py:356} INFO - k8s: setting state of ('example_python_operator_from_git_sync', 'sleep_for_5', datetime.datetime(2017, 9, 8, 14, 8)) to None
[2017-09-08 21:17:32,634] {kubernetes_executor.py:158} INFO - Event: examplepythonoperatorfromgitsyncsleepfor5-8792299a9e164ad08e381109e5052c19 had an event of type MODIFIED
[2017-09-08 21:17:32,634] {kubernetes_executor.py:168} INFO - Event: examplepythonoperatorfromgitsyncsleepfor5-8792299a9e164ad08e381109e5052c19 Succeeded
[2017-09-08 21:17:32,643] {kubernetes_executor.py:158} INFO - Event: examplepythonoperatorfromgitsyncsleepfor5-8792299a9e164ad08e381109e5052c19 had an event of type DELETED
[2017-09-08 21:17:32,643] {kubernetes_executor.py:168} INFO - Event: examplepythonoperatorfromgitsyncsleepfor5-8792299a9e164ad08e381109e5052c19 Succeeded
[2017-09-08 21:17:32,645] {kubernetes_executor.py:348} INFO - reporting (('example_python_operator_from_git_sync', 'sleep_for_1', datetime.datetime(2017, 9, 8, 14, 8)), None, 'examplepythonoperatorfromgitsyncsleepfor1-51d8fd1729d74b95a89d4a8ad38f418d')
[2017-09-08 21:17:32,645] {kubernetes_executor.py:356} INFO - k8s: setting state of ('example_python_operator_from_git_sync', 'sleep_for_1', datetime.datetime(2017, 9, 8, 14, 8)) to None
[2017-09-08 21:17:32,660] {kubernetes_executor.py:158} INFO - Event: examplepythonoperatorfromgitsyncsleepfor1-51d8fd1729d74b95a89d4a8ad38f418d had an event of type MODIFIED
[2017-09-08 21:17:32,662] {jobs.py:1381} INFO - Executor reports example_python_operator_from_git_sync.sleep_for_5 execution_date=2017-09-08 14:08:00 as None
[2017-09-08 21:17:32,662] {jobs.py:1381} INFO - Executor reports example_python_operator_from_git_sync.sleep_for_1 execution_date=2017-09-08 14:08:00 as None
[2017-09-08 21:17:32,663] {kubernetes_executor.py:168} INFO - Event: examplepythonoperatorfromgitsyncsleepfor1-51d8fd1729d74b95a89d4a8ad38f418d Succeeded
[2017-09-08 21:17:32,668] {kubernetes_executor.py:158} INFO - Event: examplepythonoperatorfromgitsyncsleepfor1-51d8fd1729d74b95a89d4a8ad38f418d had an event of type DELETED
[2017-09-08 21:17:32,669] {kubernetes_executor.py:168} INFO - Event: examplepythonoperatorfromgitsyncsleepfor1-51d8fd1729d74b95a89d4a8ad38f418d Succeeded
[2017-09-08 21:17:33,664] {jobs.py:1582} INFO - Heartbeating the process manager
[2017-09-08 21:17:33,664] {dag_processing.py:561} INFO - Processor for /root/airflow/dags/new_example_dag.py finished
[2017-09-08 21:17:33,668] {dag_processing.py:629} INFO - Started a process (PID: 223) to generate tasks for /root/airflow/dags/dag_that_fails.py - logging into /root/airflow/logs/scheduler/2017-09-08/dag_that_fails.py.log
[2017-09-08 21:17:33,671] {dag_processing.py:629} INFO - Started a process (PID: 224) to generate tasks for /root/airflow/dags/example_dag_git.py - logging into /root/airflow/logs/scheduler/2017-09-08/example_dag_git.py.log
[2017-09-08 21:17:33,673] {jobs.py:1618} INFO - Heartbeating the executor
[2017-09-08 21:17:33,673] {kubernetes_executor.py:238} INFO - Attempting to finish pod; pod_id: examplepythonoperatorfromgitsyncsleepfor5-8792299a9e164ad08e381109e5052c19; state: None; labels: {u'airflow-slave': '', u'execution_date': '2017-09-08T14_08_00', u'dag_id': 'example_python_operator_from_git_sync', u'task_id': 'sleep_for_5'}
[2017-09-08 21:17:33,674] {kubernetes_executor.py:241} INFO - finishing job ('example_python_operator_from_git_sync', 'sleep_for_5', datetime.datetime(2017, 9, 8, 14, 8))
[2017-09-08 21:17:33,674] {kubernetes_executor.py:238} INFO - Attempting to finish pod; pod_id: examplepythonoperatorfromgitsyncsleepfor5-8792299a9e164ad08e381109e5052c19; state: None; labels: {u'airflow-slave': '', u'execution_date': '2017-09-08T14_08_00', u'dag_id': 'example_python_operator_from_git_sync', u'task_id': 'sleep_for_5'}
[2017-09-08 21:17:33,675] {kubernetes_executor.py:241} INFO - finishing job ('example_python_operator_from_git_sync', 'sleep_for_5', datetime.datetime(2017, 9, 8, 14, 8))
[2017-09-08 21:17:33,675] {kubernetes_executor.py:238} INFO - Attempting to finish pod; pod_id: examplepythonoperatorfromgitsyncsleepfor1-51d8fd1729d74b95a89d4a8ad38f418d; state: None; labels: {u'airflow-slave': '', u'execution_date': '2017-09-08T14_08_00', u'dag_id': 'example_python_operator_from_git_sync', u'task_id': 'sleep_for_1'}
```

grantnicholas commented:
I added a fix that monitors the `process.is_alive()` return value and then restarts the process if it is ever not alive. Since it is hard to reproduce the issue, I added this fix in and I'm going to let it run all night and see what happens.

grantnicholas commented:
Found the issue with the first time the process "dies". In fact, the process actually gracefully exits because the generator in the `run` method stops returning values (and does not raise a `StopIteration` exception). I think this happens because periodically the kubernetes API server will drop long lived connections (otherwise a connection would stay open forever until the end of time once someone makes a watch request). 

See the logs below for confirmation. 
```
grant@grant-Latitude-E7450:~$ kc logs airflow-1377530605-3sxpw -c scheduler -f | grep -i -E '(error|warn)'
[2017-09-11 15:10:04,814] {kubernetes_executor.py:159} WARNING - Watcher process died gracefully: generator stopped returning values
```

For now I fixed this by wrapping the `run` method in an infinite loop so that if this occurs we start a watch again. The longer-term solution also involves doing checkpointing of the `resource_version` and then we would restart from the last known `resource_version`.
